compiler/src/iree/compiler/Codegen/Transforms/Transforms.cpp

```diff
@@ -615,12 +642,111 @@ struct FoldReshapeIntoInterfaceTensorLoad : OpRewritePattern<TensorReshapeOp> {
+/// hal.interface.binding.subspan.
+///
+/// For example, this matches the following pattern:
+///
+///   %subspan = hal.interface.binding.subspan ... :
+///       !flow.dispatch.tensor<readonly:tensor<3x3x1x96xf32>>
+///   %tensor = flow.dispatch.tensor.load %subspan :
+///       !flow.dispatch.tensor<readonly:tensor<3x3x1x96xf32>> ->
+///       tensor<3x3x1x96xf32>
+///   %0 = linalg.expand_reshape %tensor [
+///         affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>
+///       ] : tensor<3x3x1x96xf32> into tensor<864xf32>
+///
+/// And turns it into:
+///
+///   %subspan = hal.interface.binding.subspan ... :
+///       !flow.dispatch.tensor<readonly:tensor<864xf32>>
+///   %0 = flow.dispatch.tensor.load %subspan :
+///       !flow.dispatch.tensor<readonly:tensor<864xf32>> -> tensor<864xf32>
+struct FoldExpandShapeIntoInterfaceTensorLoad
+    : OpRewritePattern<tensor::ExpandShapeOp> {
+  using OpRewritePattern<tensor::ExpandShapeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExpandShapeOp reshapeOp,
+                                PatternRewriter &rewriter) const override {
+    Value reshapeSrc = reshapeOp.getSrc();
+    auto loadOp = reshapeSrc.getDefiningOp<IREE::Flow::DispatchTensorLoadOp>();
+    if (!loadOp) {
+      return failure();
+    }
+
+    // Make sure we are loading the full incoming subspan. Otherwise we cannot
+    // simply adjust the subspan's resultant type later.
+    if (!isFullSlice(loadOp, loadOp.getSourceType(), loadOp.getSourceDims())) {
+      return failure();
+    }
+
+    // In the corner case where the expand_shape is the source of a store, dont
+    // fold with the load. Instead fold with the store to reduce the
+    // dimensionality
+    if (reshapeOp->hasOneUse()) {
+      if (auto storeOp = dyn_cast<IREE::Flow::DispatchTensorStoreOp>(
+              *reshapeOp->getUsers().begin())) {
+        if (isFullSlice(storeOp, storeOp.getTargetType(),
+                        storeOp.getTargetDims())) {
+          return rewriter.notifyMatchFailure(reshapeOp,
+                                             "fold with store instead");
+        }
+      }
+    }
+
+    auto subspanOp = loadOp.getSource()
+                         .getDefiningOp<IREE::HAL::InterfaceBindingSubspanOp>();
+    if (!subspanOp)
+      return failure();
+
+    OpBuilder::InsertionGuard gaurd(rewriter);
+    rewriter.setInsertionPoint(subspanOp);
+
+    auto currDynamicDims = subspanOp.getDynamicDims();
+    auto currStaticDims = loadOp.getType().getShape();
+    auto currOfrDynamicDims =
+        mlir::getMixedValues(currStaticDims, currDynamicDims, rewriter);
+    std::optional<SmallVector<OpFoldResult>> expandedDims =
+        mlir::inferExpandShapeOutputShape(
+            rewriter, subspanOp.getLoc(), reshapeOp.getType(),
+            reshapeOp.getReassociationIndices(), currOfrDynamicDims);
+    if (!expandedDims) {
+      return reshapeOp.emitOpError("failure in expanded shape");
+    }
+
+    auto tensorAccess =
```

Comment:
nit: Spell out type

Reply:
Typically dont do that if this is a cast.